server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java

```diff
@@ -174,18 +175,20 @@ void getAliasMappers(Map<String, Mapper> mappers, Map<String, Mapper> aliasMappe
                                     );
                                     for (int i = fieldNameParts.length - 2; i >= 0; --i) {
                                         String intermediateObjectName = fieldNameParts[i];
-                                        ObjectMapper.Builder intermediate = new ObjectMapper.Builder(
+                                        ObjectMapper.Builder intermediate = intermediates.computeIfAbsent(
                                             intermediateObjectName,
-                                            ObjectMapper.Defaults.SUBOBJECTS
+                                            s -> new ObjectMapper.Builder(intermediateObjectName, ObjectMapper.Defaults.SUBOBJECTS)
                                         );
                                         intermediate.add(fieldBuilder);
                                         fieldBuilder = intermediate;
                                     }
-                                    aliasMappers.put(fieldNameParts[0], fieldBuilder.build(context));
                                 }
                             }
                         }
                     }
+                    for (var entry : intermediates.entrySet()) {
+                        aliasMappers.put(entry.getKey(), entry.getValue().build(context));
+                    }
```

Comment:
We probably want to deduplicate across different calls to this function.. maybe share the intermediates across calls and move this loop to the end, outside getAliasMappers?

Reply:
Good point. See: c3bbe3b682d45f948d76a389cec3029acb0b11e4